go/database/mpt/io/parallel_visit.go

```diff
@@ -0,0 +1,462 @@
+						}
+					}
+				}
+			}
+		}()
+	}
+
+	var err error
+	// wait for all go routines start to check for init errors
+	workersInitWg.Wait()
+	// read possible error
+	var chRead bool
+	for !chRead {
+		select {
+		case workerErr := <-workersErrorChan:
+			err = errors.Join(err, workerErr)
+		default:
+			chRead = true
+		}
+	}
+
+	// if init was not successful, return the error
+	if err != nil {
+		return err
+	}
+
+	// Perform depth-first iteration through the trie.
+	// This main thread iterates the trie on its own and
+	// provides the nodes to the visitor in dept-first order.
+	// This loop does not perform any I/O, and instead it queries
+	// the workers for the nodes.
+	// This provides a performance boost.
+	stack := []mpt.NodeId{root}
+	for len(stack) > 0 {
+		cur := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+
+		var res response
+		responsesMutex.Lock()
+		for {
+			found := false
+			res, found = responses[cur]
+			if found {
+				delete(responses, cur)
+				break
+			}
+			responsesCond.Wait()
+		}
+		responsesMutex.Unlock()
+
+		if res.err != nil {
+			err = res.err
+			break
+		}
+
+		if visitErr := visitor.Visit(res.node, mpt.NodeInfo{Id: cur}); visitErr != nil {
+			err = visitErr
+			break
+		}
+		switch node := res.node.(type) {
+		case *mpt.BranchNode:
+			children := node.GetChildren()
+			for i := len(children) - 1; i >= 0; i-- {
+				id := children[i].Id()
+				if !id.IsEmpty() {
+					stack = append(stack, id)
+				}
+			}
+		case *mpt.ExtensionNode:
+			next := node.GetNext()
+			stack = append(stack, next.Id())
+		case *mpt.AccountNode:
+			if !pruneStorage {
+				storage := node.GetStorage()
+				id := storage.Id()
+				if !id.IsEmpty() {
+					stack = append(stack, id)
+				}
+			}
+		}
+	}
+
+	// wait until all workers are done to read errors
+	done.Store(true)
+	workersDoneWg.Wait()
+	close(workersErrorChan)
+	for workerErr := range workersErrorChan {
+		err = errors.Join(err, workerErr)
+	}
+
+	return err
+}
+
+// position addresses a node within a tree by listing the path from the root node to the respective node.
+// Each position holds the distance of the node from the root, and an index of a branch the node
+// is attached to. For a non-branch parent node, the index is always 0.
+type position struct {
+	parent *position
+	pos    byte
+	len    byte
+}
+
+// newPosition creates a new position from a byte slice.
+func newPosition(pos []byte) *position {
+	var res *position
+	for i, step := range pos {
+		res = &position{
+			parent: res,
+			pos:    step,
+			len:    byte(i),
+		}
+	}
+	return res
+}
+
+func (p *position) String() string {
+	if p == nil {
+		return ""
+	}
+	if p.parent == nil {
+		return fmt.Sprintf("%d", p.pos)
+	}
+	return fmt.Sprintf("%s.%d", p.parent.String(), p.pos)
+}
+
+// Child creates a new position that is a child of the current position.
+// The new position is one step deeper than the current one.
+func (p *position) Child(step byte) *position {
```

Comment:
nit: this was my mistake already in the prototype -- I guess like all other members of the position, this one should also be  private

Reply:
ok, changed